docs/07-templated-components.md

```diff
@@ -1 +1,246 @@
+    {
+        if (Pizza.Toppings.Find(pt => pt.Topping == topping) == null)
+        {
+            Pizza.Toppings.Add(new PizzaTopping() { Topping = topping });
+        }
+    }
+
+    void RemoveTopping(Topping topping)
+    {
+        Pizza.Toppings.RemoveAll(pt => pt.Topping == topping);
+        StateHasChanged();
+    }
+}
+```
+
+Now we can remove `@inject OrderState OrderState` from the top of `ConfigurePizzaDialog.cshtml`. 
+
+Lastly, update the rest of this code to use the parameters and remove the lingering usage of `OrderState` from `ConfigurePizzaDialog.cshtml`. At this point the code should compile, but it will not run correctly because we haven't updated `Index.cshtml` yet.
+
+----
+
+Recall that our `TemplatedDialog` contains a few `div`s. Well, this duplicates some of the structure of `ConfigurePizzaDialog`. Let's clean that up. Open `ConfigurePizzaDialog.cshtml`; it currently looks like:
+```html
+<div class="dialog-container">
+    <div class="dialog">
+        <div class="dialog-title">
+        ...
+        </div>
+        <form class="dialog-body">
+        ...
+        </form>
+
+        <div class="dialog-buttons">
+        ...
+        </div>
+    </div>
+</div>
+```
+
+We should remove the outermost two layers of `div` elements since those are now part of the `TemplatedDialog` component. After removing these it should look more like:
+```html
+<div class="dialog-title">
+...
+</div>
+<form class="dialog-body">
+...
+</form>
+
+<div class="dialog-buttons">
+...
+</div>
+```
+
+## Using the new dialog
+
+We'll use this new templated component from `Index.cshtml`. Open the `Index.cshtml` and find the block of code that looks like:
+```
+@if (OrderState.ShowingConfigureDialog)
+{
+    <ConfigurePizzaDialog />
+}
+```
+
+We are going to remove this and replace it with an invocation of the new component. Replace the block above with code like the following:
+```
+<TemplatedDialog Show="OrderState.ShowingConfigureDialog">
+    <ConfigurePizzaDialog 
+        Pizza="OrderState.ConfiguringPizza" 
+        OnCancel="OrderState.CancelConfigurePizzaDialog" 
+        OnConfirm="OrderState.ConfirmConfigurePizzaDialog" />
+</TemplatedDialog>
+```
+
+This is wiring up our new `TemplatedDialog` component to show and hide itself based on `OrderState.ShowingConfigureDialog`. Also, we're passing in some content to the `ChildContent` paramter. Since we called the parameter `ChildContent` any content that is placed inside the `<TemplatedDialog> </TemplatedDialog>` will be captured by a `RenderFragment` delegate and passed to `TemplatedDialog`. 
+
+note: A templated component may have multiple `RenderFragment` parameters, what we're showing here is a convenient convention when the caller wants to provide a single `RenderFragment` that represents the *main* content.
+
+At this point it should be possible to run the code and see that the new dialog works correctly. Verify that this is working correctly before moving on to the next step.
+
```

Comment:
This is great.

I wonder if there's any opportunity to reuse the `TemplatedDialog` just to get the feel-good effect of making another dialog so easily. Perhaps we could have an `OrderWasPlaced` dialog after orders are placed, which simply says "Thanks for your order [OK]", or a "delivery completed" overlay we show on orders that are in "Delivered" status.

Reply:
I think we should consider doing this in a later revision. It would make this more satisfying, but adding the additional steps to create another use case for the dialog would make this too long.

After adding the templated dialog and templated list this section might be our longest.